### DISCUSSION TITLE
Valtio v2 Roadmap

### DISCUSSION THREAD

**Discussion by dai-shi** [pmndrs/valtio#703] [Ideas]
(This description will be updated over time. Please quote sentences when you add comments.)

## Why do we need v2?

When we released v1, we had some assumption about supporting suspense.
However, as [the use RFC](https://github.com/reactjs/rfcs/pull/229) implies, the assumption wasn't correct. (The library can't throw promises internally without hooks.)

We need a new breaking API to follow the proposed way.

## When will v2 be released?

It will be after the RFC is finalized. We may prepare some pre-releases.
(If it takes too long for the RFC to be finalized, we will re-consider.)

## What will be breaking?

Unfortunately, APIs are not fully compatible with v1. (Some existing tests will fail.)

In v2, Valtio doesn't resolve promises, and users have to explicitly use `use` to unwrap promises.

```js
// v1
import { proxy, useSnapshot } from 'valtio'

const state = proxy({ data: fetch(...).then((res) => res.json()) })

const Component = () => {
  const snap = useSnapshot(state)
  return <>{JSON.stringify(snap.data)}</>
}
```

```js
// v2
import { use } from 'react'
import { proxy, useSnapshot } from 'valtio'

const state = proxy({ data: fetch(...).then((res) => res.json()) })

const Component = () => {
  const snap = useSnapshot(state)
  return <>{JSON.stringify(use(snap.data))}</>
  // If `data` is not an object, you can directly embed it in JSX.
  // return <>{snap.data}</>
}
```

### React version

v2 will require at least React 18.
This will allow to drop `use-sync-external-store` package. 
https://github.com/pmndrs/valtio/issues/698#issuecomment-1477763683

### Use of initial object (copying behavior in v1)

v1 copies initial object that is passed to `proxy()` (even internally).
This is a big design decision to avoid confusion. It has been working well for most cases.
But, some people reported edge cases that the copying behavior is more confusing.
As this behavior is really hard to disable, we make it opt-in in v2.

```js
// v1
import { proxy } from 'valtio';

const state = proxy({ count: 1, obj: { text: 'hi' } })
state.obj = { text: 'hello' }
```

To get the same behavior in v2:
```js
// v2
import { proxy } from 'valtio';
import { deepClone } from 'valtio/utils';

const state = proxy(deepClone({ count: 1, obj: { text: 'hi' } }))
state.obj = deepClone({ text: 'hello' })
```

But, that's a rare case. Unless you re-use the initial object, which is not recommended (the strong recommendation is to throw it right away), you don't need `deepClone`.

### Deprecated features

All features marked as deprecated in v1 will be removed in v2.

### Possible better snapshot cache invalidation

#712, but I'm not sure if it's resolvable.

### `derive-valtio` will be removed from the dependency

It has to be manually installed.

**Comment by Aslemammad** [comment]
The use of the initial object is more reasonable I believe since deepClones may be unexpected for some users and they may not prefer it! While for the rest, using the initial object is not something unexpected and weird! So big up for this one. Thank you.

**Comment by lovetingyuan** [comment]
why not offer option for proxy
`const state = proxy(deepClone({ count: 1, obj: { text: 'hi' } }))` => `const state = proxy({ count: 1, obj: { text: 'hi' } }, { deepClone: true })`

**Reply by dai-shi** [reply]
It's very rare that `deepClone` is required. The recommendation is not to keep the original object.
Also, we generally try to avoid options in API.

**Reply by jacob-orbiit** [reply]
I think the option would be better, simpler DX (Principle of Least Knowledge). Personally, I find nesting functions cumbersome, as something that could/should be abstracted away. It also means fewer implementation details exposed to consumers.

**Reply by dai-shi** [reply]
Personally, I feel nesting functions better, so it's a different opinion.
But, the point is using `deepClone` is almost unnecessary and not recommended for typical usage. Having an option means people has to understand when to use and when to not use. I'm fine to not expose `deepClone` if it's better DX.

**Reply by jacob-orbiit** [reply]
Wouldn't they need to understand it regardless? Exposing it as its own function means it can be erroneously used pretty much anywhere. As an option, it's only indirectly exposed to a user where it may be relevant (and otherwise not available), so the surface-area of misuse is far smaller. I think it would be better documentation-wise too because it would only be documented where it could be relevant.

**Reply by dai-shi** [reply]
I think our opinions simply conflict.
Another reason I prefer add-ons (as oppose to options) is to make the main bundle size small.

**Reply by pastelmind** [reply]
Would `deepClone()` apply to objects that are added later to the state tree?

```js
// Hypothetical valtio v2
const obj = {}
const p1 = proxy(obj) // Wraps original object

const child1 = { age: 1 }
p1.child1 = child1 // Also sets obj.child1
console.log(obj.child1) // { age: 1 }

// p1.child1 wraps child1
p1.child1.age = 10
console.log(child1.age) // 10

const p2 = proxy(deepClone(obj)) // Wraps deep clone
const child2 = { age: 2 }
p2.child2 = child2 // Will this be deep-cloned, too?

p2.child2.age = 20
console.log(child2.age) // ?
```

**Reply by dai-shi** [reply]
No, users have to do it by themselves.
```diff
- p2.child2 = child2 // Will this be deep-cloned, too?
+ p2.child2 = deepClone(child2) // It's up to developers whether to use deepClone or not.
```

**Reply by dai-shi** [reply]
That said, we would never recommend using the original object. We should assign it and throw it away.
```js
const p1 = proxy({ c: 0 }); // do this

const obj = { c: 0 }; const p2 = proxy(obj); // avoid this

p1.child = { foo: 'bar' }; // do this

const child = { foo: 'bar' }; op1.child = child; // avoid this
```
In that way, the behavior change between v1 and v2 would not be noticeable.

**Reply by pastelmind** [reply]
> we would never recommend using the original object

Why so?

**Reply by dai-shi** [reply]
Because there exists two objects with the same value with different capability, which I think is super confusing.

**Reply by ScreamZ** [reply]
I agree, I'm not sure it's written explicitly it the documentation, but this advice of throwing it should be marked as a big highlight because it can cause big issues isn't ? Maybe for v2 doc

**Reply by dai-shi** [reply]
Yeah, with v1, it's fairly okay, because `proxy()` is "pure". Let's highlight it in v2 docs. By the way, would you be interested in contributing or review for the v2 docs?

**Reply by ScreamZ** [reply]
I love documenting things, interested in both. I'm currently mainly using valtio on server side to power state synchronization mechanism over websocket for an escape room solution driving IoT devices (like doors lockers, lights, dmx, …). Also with react, as my code heavily relies on valtio I would be pleased to contribute back. And I'm an absolute fan of @pmndrs content ahah

**Reply by dai-shi** [reply]
You can open a new discussion about v2 docs and draft some contents, or you can directly open a PR to `v2` branch, where we can discuss too. Thanks for your contribution!

**Reply by dai-shi** [reply]
btw, we can technically release Valtio v2.0.0 right after React 19 is released. And, what's blocking then is v2 docs. So, contributions are very welcome.

**Reply by ScreamZ** [reply]
I'll work on a PR on V2 in the next days

**Reply by dai-shi** [reply]
@ScreamZ Just curious if you are still interested in it.

**Reply by ScreamZ** [reply]
yes, I'm on it :)

**Comment by texastoland** [comment]
> In v2, Valtio doesn't resolve promises, and users have to explicitly use use to unwrap promises.

I think your solution is nearly as nice taking advantage of React's `status`, `value`, and `reason` props even outside Suspense!

1. Any reason not to set `status = 'pending'` initially?
    https://github.com/pmndrs/valtio/blob/9e6868c8e60f6c0057cb2d314e0f4cff22d42b8e/src/vanilla.ts#L273-L284

2. I'm not sure but I think you could also pass `use` to `snapshot` implicitly:
    ```ts
    type PromiseResult<Value> = 
      | { status: 'pending' }
      | { status: 'fulfilled', value: Value }
      | { status: 'rejected', reason: unknown }

    // in a React utility
    const suspend = (target) => use(snapshot(target))

    // infer PromiseResult<Data>
    const result = suspend(proxy.api.promise)
    ```

**Reply by dai-shi** [reply]
> 1. Any reason not to set `status = 'pending'` initially?

Good point. I think I just forgot it. But, please remember that this code will be removed in v2.

> 2. I'm not sure but I think you could also pass `use` to `snapshot` implicitly:

Yeah, but then it depends on React, while `snapshot` is pure JS function. And again, this feature will be removed in v2.

**Reply by texastoland** [reply]
> Yeah, but then it depends on React, while snapshot is pure JS function. And again, this feature will be removed in v2.

Correct it would be a single line utility imported in https://github.com/pmndrs/valtio/blob/main/src/react/utils.ts.

> But, please remember that this code will be removed in v2.

I'm confused why this specific code would be removed? The only React behavior that's changing is throwing promises. That part is limited to `defaultHandlePromise` which should become optional instead:
https://github.com/pmndrs/valtio/blob/9e6868c8e60f6c0057cb2d314e0f4cff22d42b8e/src/vanilla.ts#L89-L104

By returning the promise object with `status`, `value`, and `reason` props you'd be providing DX like most query libraries (compared to current behavior):

```js
const { status, value: data, reason: error } = snapshot(proxy.api.promise)

if (status === 'rejected') {
  // with the type from my snippet
  // TypeScript will even know
  // data is undefined
  // and error isn't!
}
```

The Valtio implementation in my previous comment is (not accidentally?) exactly what React does internally:
https://github.com/facebook/react/blob/fdb368d9e7430b50b9c506e76ccd46bf0576ef9b/packages/react-reconciler/src/ReactFiberThenable.js#L136-L151

I was investigating your implementation for Legend. It's similar to Valtio without the read/write separation. Here's my full writeup about `use`:
https://github.com/LegendApp/legend-state/issues/162#issuecomment-1653694337

**Reply by dai-shi** [reply]
React will be forcing to use `use` in the future. If you throw a promise without using `use`, React will warn you.
So, Valtio v2 will completely change. It will no longer do anything with promises, to follow the React recommendation. Valtio users have to explicitly use `use`.
`snapshot` in Valtio is a pure JS function which has nothing to do with React hooks, so we can't/shouldn't use `use` internally.

**Reply by texastoland** [reply]
@dai-shi We're saying the same thing 🤓

The code I originally linked is where you set `status` on promise objects. If you return these enhanced promises in v2 you'll provide better DX. React also uses these exact properties to optimize `use` internally. Are you planning to remove `status` too?

I'm not suggesting you change `snapshot` except to remove `defaultHandlePromise` which throws promises. You could export a React utility (similar to `useProxy`) to not have to wrap `snapshot` in `use`:

```js
const { value } = use(snapshot(target))
```

I'll leave it at that. Love your work!

**Reply by dai-shi** [reply]
> If you return these enhanced promises in v2 you'll provide better DX.

Ah, I see what you mean. I don't think it will change DX, but it will be a little more efficient so theoretically better UX.
But, I'll first probably remove it. It helps to reduce code which is also very important.

BTW, Jotai return those enhanced promises. So, I'd say they are slightly different design decisions.

**Reply by texastoland** [reply]
I previously read part of the source for Jotai too. It's beautiful (much simpler as you've said on Twitter) 😲 I love the ecosystem you've built around it too. But I use Vue whenever possible and and prefer the more "magical" API.

**Comment by dai-shi** [comment]
#810 is work in progress.
We will release alpha version first.

**Comment by hrsh7th** [comment]
I suggest customizing promises so that users can choose between two common types of state handling.

1. `use(snapshot.fetched)`
2. `const { status, data } = snapshot.fetched`

What do you think about this?

**Reply by dai-shi** [reply]
We generally don't deal with promises and especially v2 goes further away.

I think loading status is out of the scope of Valtio itself, but it might be interesting to have a solution as a third party library in the ecosystem.

**Reply by dai-shi** [reply]
Actually, Valtio follows React convention, so properties such as status and value are available.

**Comment by ayoubkhan558** [comment]
I just wanted to express my immense appreciation for the outstanding work you've done with Valtio Version 1. It's so seamless that I often forget I'm using an external library.
As I eagerly await Version 2, I'm curious to see if you'll maintain the same seamless experience or introduce any changes. ✨

**Reply by dai-shi** [reply]
Thanks.

I highly recommend you to test v2-beta right now. #810 

> https://twitter.com/dai_shi/status/1786556901507105009

As noted, unless we receive any feedback, v2 will be finalized.

**Comment by dai-shi** [comment]
#810 is merged.
